tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -416,6 +416,105 @@ bool ROCMBlas::DoBlasInternalImpl(FuncT rocblas_func, Stream *stream,
   return ret == rocblas_status_success;
 }
 
+
+template <class T>
+void SplitComplex(T* as, const T* a, int n, void* stream, bool conjugate);
+template <class T>
+void CombineComplex(T* a, const T* as, int n, void* stream);
+template <class T, int N>
+void AddComplex(T* as, const T* a, int n, void* stream);
+
+template <typename FuncT, typename T, typename... Args> 
+port::Status ROCMBlas::DoBlasComplex(ScratchAllocator* scratch_allocator,
+    Stream* stream,
+    const T* a, const T* b, T* c,
+    uint64_t na, uint64_t nb, uint64_t nc, 
+    std::complex<T> alpha, std::complex<T> beta,
+    FuncT rocblas_func,
+    int conj_op,
+    Args... args)
+{
+    if(alpha.imag()!=0.0 || beta.imag()!=0.0)
+      return port::Status(
+          port::error::INTERNAL, "ROCMBlas only supports real-valued alpha & beta");
+    auto hs = AsGpuStreamValue(stream);
+    auto sz_alloc = na * sizeof(T) * 2 + nb *sizeof(T)*2 + nc*sizeof(T)*4;
+    DeviceMemory<uint8_t> scratch;
+    std::unique_ptr<TemporaryDeviceMemory<uint8_t> > temp_memory;
+    if(scratch_allocator != nullptr)
+    {
+      SE_ASSIGN_OR_RETURN(scratch, scratch_allocator->AllocateBytes(sz_alloc));
+    }
+    else
+    {
+      SE_ASSIGN_OR_RETURN(temp_memory, stream->AllocateTemporaryArray<uint8_t>(sz_alloc));
+      scratch =DeviceMemory<uint8_t>(*(temp_memory->mutable_device_memory()));
+    }
+    uint8_t* p_scratch = GpuMemoryMutable(&scratch);
+
+    T* pas = reinterpret_cast<T*>(p_scratch);
+    T* pbs = reinterpret_cast<T*>(p_scratch + na * sizeof(T) * 2);
+    T* pcs = reinterpret_cast<T*>(p_scratch + na * sizeof(T) * 2 + nb*sizeof(T)*2);
+    SplitComplex(pas, a, na, hs, (conj_op & 1));
+    SplitComplex(pbs, b, nb, hs, (conj_op & 2));
+    SplitComplex(pcs, c, nc, hs, false);
+    hipMemsetAsync(pcs+nc*2, 0, nc*2*sizeof(T), hs);
+    bool ok = true;
+    for(int i=0; i<4; i++)
+      ok = ok && DoBlasCall(pas + ((i&1) ? na : 0), pbs + ((i&2) ? nb : 0), pcs+i*nc, (const T*)&alpha, (const T*)&beta, rocblas_func, stream, args...);
+    if(!ok)
+      return port::Status(
+        port::error::INTERNAL, "DoBlasInternal failure");
+    AddComplex<T, 1>(pcs + nc, pcs + 2*nc, nc, hs);
+    AddComplex<T,-1>(pcs, pcs + 3*nc, nc, hs);
+    CombineComplex(c, pcs, nc, hs);
```

Comment:
I think this is the core of it, very clever! I kind of get the general idea of what you are trying to do here 1. split the original complex data into a longer buffer; 2. Call BLAS function in the extended buffer 3. Re-consolidate the long buffer back to original complex format. Though I have trouble understanding all the details here.

Could you start a comment block specifying all the details, ie, what is complex number layout before & after; why are you splitting it in the specific way; why is this way mathematically equivalent to computing complex type in place.